### ISSUE TITLE
Merge different tracks to one

### ISSUE THREAD

**Issue by photastic** [mapillary/mapillary_issues#122]
Is there a function planned to merge different tracks with one?
This could be applied if you recorded a certain street several times
in the same direction.
It would make navigation a lot easier because you won't have to change to a different track when you want to see a picture that lays between two of the current track.

Regards photastic

**Comment by pyrog** [comment]
@photastic :+1: 

@gyllen It could be easier if Mapillary use **map matching** algorithm on tracks: https://github.com/mapillary/mapillary_issues/issues/224#issuecomment-95232924

**Comment by ulikoehler** [comment]
I'm currently running into this issue as I'm uploading tons of High-res images (2.0 MB / img) using a Raspberry-Pi based setup 24/7 on my ADSL line.  Due to the DSL reconnecting once every night and due to bugs in the upload-without-tags script, I tend to end up with a lot of split sequences after following the *if the upload fails mid-sequence* procedure.

This doesnt't mean "a sequence split at some point into two consecutive sequences" but mostly "one sequence with hundred images, plus a 10-image sequence which basically follows the same track as the longer sequence". The images from the short sequence are missing from the long sequence and are no duplicates.

I can provide examples if that would help.

**Comment by gyllen** [comment]
@ulikoehler An example would greatly help

**Comment by ulikoehler** [comment]
@gyllen 
One example would be
https://www.mapillary.com/app/?lat=47.869614678604364&lng=10.177116260517266&z=17.647011480849454&pKey=qZSg-YFIk2VWiiSYeZDSEw&focus=map

At the bridge, you can see multiple sequences, although I walked over it only once. If they would be merged, the sequence would be OK.

Another example is directly to the east/southeast of the bridge where you can see multiple sequences "in parallel"

Basically you can identify it by "a sequence of many pictures being intersected by almost parallel lines with only a few pictures".

**Comment by gyllen** [comment]
Were these images uploaded in the same sequence https://www.mapillary.com/app/?lat=47.874744722222204&lng=10.177490000000034&z=17&pKey=WM0SvpFG_bVAiRHxtjs3zg&focus=photo ?

**Comment by ulikoehler** [comment]
@gyllen If you mean the two sequences that are visible (at your link) in the lower left corner, these were actually different sequences (i.e. I stopped recording in the app manually and started again after stopping for a few seconds).

As for many other sequences, I uploaded them using my BeagleBoard (like a Raspberry Pi) overnight using mapillary_tools `upload_with_authentication.py`, so while they are part of the same source sequence (i.e. folder generated by the app), they might have been split into different sequences during upload due to ADSL reconnection, some bugs in the upload script etc.

For example, the two sequences here:
https://www.mapillary.com/app/?lat=47.87365841078608&lng=10.17776430548389&z=17.34866947996767&pKey=3BHplmUr4fLmXWt0zlyROg&focus=map
were part of the same sequence (as taken by the app), and I assume they have been split into two "parallel" sequences by the upload process.

**Comment by gyllen** [comment]
That sequence actually looks more like an error in the uploader. Did you upload them to the same directory and _then_ applied the done command?

**Comment by ulikoehler** [comment]
I don't have records for the specific sequences, but in general it seems like `upload_with_authentication.py` is related to the issue.

I have to separate DSL connections (at different places). At one of them, I've placed my BeagleBoard to upload basically 24/7. I use it with `auto_done` This *mostly* works really well, except for it sometimes being stuck at "0 images left" forever, in which case I restart (usually it uploads one image after that). Also, I'm using git rev `0f84a7a6e628e33cf84d308d913b6d64edec329b` as the `master` branch currently seems to be stuck more often (i.e. more restarts, i.e. more different sequences).

The other usecase is uploading on my notebook or desktop when I'm using it. When I'm shutting it down, the sequence upload will be aborted. I regularly go to https://legacy.mapillary.com/map/upload/im and publish the remaining sequences.

I also upload a few (short) sequences from my smartphone, but I'm pretty much limited by free space on the flash (~4.4 GB max, ~2k images) and I'm copying all the image folders to a USB stick (via OTG adapter) and then move them either to the notebook or to the beagleboard, depending on the location. I generally do uploads in a FIFO manner, although I sometimes upload small folders first to get rid of them.

I think the uploader *might* be bugged in two ways:
1. It seems to me like its not uploading images in order, i.e. I expect that an aborted sequence will cause two distinct (non-overlapping in time) sequences to appear on Mapillary
2. The script should save the sequence ID in the image folder so it continues to upload to the same sequence after being restarted (e.g. after computer reboot etc). Not sure though if this has unintended side-effects.

I'm pretty sure this specific bug can be fixed by improving the uploader script. However, there currently seem to be many ways to unintentionally break a sequence during upload to mapillary - so having an automated or manual sequence merging feature in the future will work around the issue as well.

**Comment by gyllen** [comment]
Have you tried `upload_with_preprocessing.py`, its a more stable script. Could you try to change and then we can see if the error still exists.

ALSO this beagleboard setup sounds really cool. Do you have anything to share around that project?

**Comment by ulikoehler** [comment]
@gyllen  Sorry for the delay in responding.

I've now tested `upload_with_preprocessing.py`, which indeed seems to fix the issue that multiple sequences are generated. Based on the CLI output and the result (as seen on the webinterface), I don't think there should be any issue with that behaviour any more.

I have no practical idea yet on how to deal with existing sequences, but it's not an urgent issue whatsoever.

I've noticed that using my ADSL connection, the upload script is rather unstable, i.e. it doesn't always restart after certain errors (e.g. DSL reconnect in the middle of the night). Therefore I've modified my upload process (beagleboard is currently down due to some SD card issue, so I'm running the same script on my desktop currently).

What changed is basically that the script does not any more upload the imgs to mapillary directly, but instead rsync it to my VPS, which has a stable 200M internet connection. Although the upload seems to be rate-limited to ~400 kbyte/s (doesn't matter to me, it's running 24/7 anyway), uploading is 100% stable from the VPS whereas it caused issue in the past. Also, using `rsync` I can use `--remove-source-files`, which allows me to delete local files before a full sequence has been uploaded, freeing more space on my USB stick to upload more mapillary images.

Furthermore, `rsync` allows me to use `--bwlimit`. The raw mapillary uploader, as well as the App (with good wifi connection), slows down the entire DSL line because the upload bandwith is occupied by mapillary. As its all HTTPs, I can't really do any QoS prioritizing on my router. I have ~75 kbytes/s upload bandwidth. Experiments have shown that 40 kbps "background upload" bandwidth have no effect on other services & IPTV whatsoever, so I'm using this setting.

Basically this script does a number of different things
* Copy GPX files to a local folder, because it's lost once the data is uploaded to mapillary (not sure if there is any significant use for it, though)
* Upload **smallest mapillary folder first**, because this allows me to get as many photos as possible published within any timeframe.
* After a folder upload is finished, it creates an empty `<foldername>-finished.txt` text file, which shows me on the server what directories I can upload.
* Remove local directories which have already been uploaded completely
* Do everything in a loop until no sequence folders are left

Here's the script:
```bash
#!/bin/sh
while true ; do # Retry uploads indefinitely
  # Copy GPX
  cp $(find . -name *.gpx) ~/gpx
  # Break if there are no folders
  ls 201* >/dev/null 2>/dev/null
  if [ $? -gt 0 ]; then echo "Finished!"; break ; fi
  # rsync --remove-source-files does not remove directories, so we do it now.
  # This might only remove some nested directories, but it doesnt matter if it takes multiple
  # runs to actually remove the directory
  find . -type d -print0 | xargs -0 rmdir 2> /dev/null
  # Upload smallest folders first (reduces total folder size fastest)
  for i in $(du -sb 201* | sort -n | cut -f2) ; do # Upload smaller folders first.
	rsync -Pavz --bwlimit=40 --remove-source-files "$i" myserver.domain:mapillary && ssh myserver.domain bash -c "echo ! > ~/mapillary/${i}-finished.txt"
  done
  # Wait between retries. Prevents 100% CPU load in case of network failure etc.
  sleep 5
done
```

The script on the server is simply a shell script setting the ENV vars for the upload script:

```bash
#!/bin/sh
MAPILLARY_SIGNATURE_HASH="xxx=" MAPILLARY_EMAIL="xxx" MAPILLARY_PASSWORD="xxx" MAPILLARY_USERNAME=xxx MAPILLARY_PERMISSION_HASH="xxxx==" ./upload_with_preprocessing.py --auto_done $1
```
On the server, I run (in tmux) `./do-upload.sh ~/mapillary/2017_06_10_16_49_29_279` and after that finishes, `rm -rf ~/mapillary/2017_06_10_16_49_29_279`. I select

**Comment by gyllen** [comment]
Check out the new mapillary_tools, they are better optimized https://github.com/mapillary/mapillary_tools

**Comment by ulikoehler** [comment]
@gyllen Yes, I didnt encounter any of those issues with recent versions of these tools. They work well for me.
